### PR TITLE
Fix: make loading faster

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/utils/dag.py [hunk 1]
```diff
@@ -80,23 +83,25 @@ def graph(self) -> t.Dict[T, t.Set[T]]:
...
-        result: t.List[T] = []
+        if self._sorted is None:
+            self._sorted = []
 
-        unprocessed_nodes = self.graph
-        while unprocessed_nodes:
-            next_nodes = {node for node, deps in unprocessed_nodes.items() if not deps}
+            unprocessed_nodes = self.graph
+            while unprocessed_nodes:
+                next_nodes = {node for node, deps in unprocessed_nodes.items() if not deps}
```
Comment (crericha) [inline sqlmesh/utils/dag.py:94]: We should raise an exception if next_nodes is empty. It means we have a circular dependency between the remaining unprocessed nodes.

File: sqlmesh/core/context.py [hunk 2]
```diff
@@ -829,7 +828,7 @@ def get_dag(self, format: str = "svg") -> graphviz.Digraph:
 
         graph = graphviz.Digraph(node_attr={"shape": "box"}, format=format)
 
-        for name, upstream in self.dag.graph.items():
+        for name, upstream in self.dag._graph.items():
```
Comment (izeigerman) [inline sqlmesh/core/context.py:831]: Why did we make `_graph` private and are now accessing it from the outside?

File: sqlmesh/core/loader.py [hunk 3]
```diff
@@ -41,21 +41,22 @@
...
-    for name in dag.sorted():
+    for name in dag.sorted:
         model = models.get(name)
 
         # External models don't exist in the context, so we need to skip them
         if not model:
             continue
 
         model.update_schema(schema)
+        # don't add columns to types if empty?
```
Comment (izeigerman) [inline sqlmesh/core/loader.py:52]: Is this a FIXME / TODO?

File: sqlmesh/utils/cron.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import typing as t
+
+from croniter import croniter
+
+from sqlmesh.utils import ttl_cache
+from sqlmesh.utils.date import TimeLike, now, to_datetime
+
+
+@ttl_cache(maxsize=1280000)
```
Comment (izeigerman) [inline sqlmesh/utils/cron.py:11]: what does this magic number mean? Do we need a named constant?

File: sqlmesh/core/context.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/core/context.py:831]: because by default graph does a copy. should i make it not copy by default?

File: sqlmesh/core/model/meta.py [hunk 5]
```diff
@@ -265,10 +267,10 @@ def normalized_cron(self) -> str:
             return "0 0 * * *"
         return ""
 
-    def croniter(self, value: TimeLike) -> croniter:
+    def croniter(self, value: TimeLike) -> CroniterCache:
         if self._croniter is None:
-            self._croniter = croniter(self.normalized_cron())
-        self._croniter.set_current(to_datetime(value))
+            self._croniter = CroniterCache(self.normalized_cron())
+        self._croniter.curr = value
```
Comment (izeigerman) [inline sqlmesh/core/model/meta.py:273]: Why are we not passing this in constructor? We basically waste another `now` call.

Comment (tobymao) [reply sqlmesh/core/model/meta.py:273]: what do you mean?

Comment (tobymao) [reply sqlmesh/core/model/meta.py:273]: ah you mean value?

File: sqlmesh/utils/cron.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+def cron_next(cron: str, time: TimeLike) -> float:
+    return croniter(cron, to_datetime(time)).get_next()
+
+
+@ttl_cache(maxsize=1280000)
+def cron_prev(cron: str, time: TimeLike) -> float:
+    return croniter(cron, to_datetime(time)).get_prev()
```
Comment (izeigerman) [inline sqlmesh/utils/cron.py:18]: why aren't we better off caching the `croniter` instance itself?

File: sqlmesh/utils/date.py [hunk 7]
```diff
@@ -22,24 +24,33 @@
     from sqlmesh.core.scheduler import Interval
 
 
-def now() -> datetime:
+def now(floor: bool = True) -> datetime:
```
Comment (izeigerman) [inline sqlmesh/utils/date.py:27]: bad argument name. Suggestions: `round_to_minute`.

File: sqlmesh/utils/date.py [hunk 8]
```diff
@@ -22,24 +24,33 @@
...
     """
-    return datetime.utcnow().replace(tzinfo=UTC)
+    now = datetime.utcnow()
+    if floor:
+        return now.replace(second=0, microsecond=0, tzinfo=UTC)
+    return now.replace(tzinfo=UTC)
 
 
-def now_timestamp() -> int:
+def now_timestamp(floor: bool = False) -> int:
```
Comment (izeigerman) [inline sqlmesh/utils/date.py:43]: ditto

File: sqlmesh/utils/date.py [hunk 9]
```diff
@@ -96,6 +107,7 @@ def to_timestamp(value: TimeLike, relative_base: t.Optional[datetime] = None) ->
     return int(to_datetime(value, relative_base=relative_base).timestamp() * 1000)
 
 
+@ttl_cache(maxsize=1280000)
```
Comment (izeigerman) [inline sqlmesh/utils/date.py:110]: ditto: magic number. Should this be default in the `ttl_cache`?

File: sqlmesh/utils/cron.py (hunk 6 above)
Comment (tobymao) [reply sqlmesh/utils/cron.py:18]: cache works on methods, also ceo inter is stateful

File: tests/utils/test_yaml.py [hunk 10]
```diff
@@ -13,7 +13,6 @@ def test_yaml() -> None:
     prod:
       type: postgres
       host: 127.0.0.1
-      # IMPORTANT: Make sure to quote the entire Jinja string here
```
Comment (izeigerman) [inline tests/utils/test_yaml.py:16]: why was this removed?

File: sqlmesh/utils/date.py (hunk 7 above)
Comment (tobymao) [reply sqlmesh/utils/date.py:27]: https://pandas.pydata.org/docs/reference/api/pandas.Timestamp.floor.html

File: sqlmesh/utils/date.py (hunk 9 above)
Comment (tobymao) [reply sqlmesh/utils/date.py:110]: there already is one, ttl_cache is time based, maxsize is pass through to the python cache method

File: tests/utils/test_yaml.py (hunk 10 above)
Comment (tobymao) [reply tests/utils/test_yaml.py:16]: because the C yaml parser doesn't preserve comments

File: sqlmesh/utils/date.py (hunk 7 above)
Comment (izeigerman) [reply sqlmesh/utils/date.py:27]: well in that case, there's an additional argument which specifies resolution. We don't have it there. The core of my concern here is that it's unclear to which granularity it will be floored.

File: sqlmesh/utils/date.py (hunk 9 above)
Comment (izeigerman) [reply sqlmesh/utils/date.py:110]: the current default is 60. Why have it if we override it everywhere. Should the default just be 1280000?

File: sqlmesh/core/model/meta.py (hunk 5 above)
Comment (izeigerman) [reply sqlmesh/core/model/meta.py:273]: yes, exactly

**Review by izeigerman — APPROVED** [review]
